### PR TITLE
Grouper refactor: changes to grouper to follow modular best practices

### PR DESCRIPTION
**Pull request by delarocamcas**
Files changed:
 Grouper.py

Explanation of changes:
Greetings, the goal of this PR is to move every line of code into a commented and named function. This way the order of steps and code can be understood in a cleaner and easier manner. I was very careful to do this manually instead of using AI and did not change any of the underlying code, other than moving it into sensible functions based on your in-line comments. 

points:
- This should be callable in the exact same way as the original file
- I did not add any of my custom code from the earlier version, this is 100% original grouper code.
-  this included your most recent changes made over the past month

### CODE REVIEW

File: Grouper.py [hunk 1]
```diff
@@ -66,12 +68,7 @@ def preprocess(text):
     text = re.sub(r'\bkm\.?\b', ' kilometers ', text, flags=re.IGNORECASE)
     text = re.sub(r"(\d+)\s*['’]", r"\1 feet", text)
 
-     # change " m " to meters if preceding number is above 20, miles if below 20
```
Comment (delarocamcas) [inline Grouper.py:69]: This was split off into its own standalone function. I am not sure about you. But I am wary of functions inside of functions

File: Grouper.py [hunk 2]
```diff
@@ -167,31 +164,6 @@ def convert_m_unit(match):
     for abbr, full in abbr_map.items():
         text = re.sub(rf'\b{abbr}\b', full, text, flags=re.IGNORECASE)
 
-    # Fuzzy-correct direction typos after normalization (Commented out cause it is more trouble than its worth.)
```
Comment (delarocamcas) [inline Grouper.py:170]: Not sure if you for-see wanting to save this (although its saved in past commits) but I removed it

File: Grouper.py [hunk 3]
```diff
@@ -270,8 +241,8 @@ def convert_m_unit(match):
             'east-northeast', 'east-southeast',
             'west-northwest', 'west-southwest'
     ]
-    dir_pattern = '|'.join(directions)
```
Comment (delarocamcas) [inline Grouper.py:273]: I did not see this line used or referenced anywhere else , so this has been removed.

File: Grouper.py [hunk 4]
```diff
@@ -387,51 +358,36 @@ def convert_m_unit(match):
...
-    if results:
-        results.sort(key=lambda x: (x[1], float(x[0])))
-        return results
+    return results
 
-    # --- Fallback logic (if no standard pattern was matched) ---
-    # Try to detect *exactly one* number+unit and *exactly one* direction in any order
 
-    # Find first occurrence of a number + unit (e.g., "9 miles")
+def fallback_direction(text):
```
Comment (delarocamcas) [inline Grouper.py:381]: Split fallback directions into their own function to shorten the direction normalization function

File: Grouper.py [hunk 5]
```diff
@@ -453,85 +409,143 @@ def extract_distance_direction(text):
...
-    exit()
+    ext = os.path.splitext(csv_path)[1].lower()
+    sep = '\t' if ext == '.tsv' else ',' if ext == '.csv' else None
 
-df = pd.read_csv(csv_path, sep=sep)
+    if sep is None:
+        print("Unsupported file type. Please provide a .csv or .tsv file.")
+        exit()
 
-grouping_field = "bels_location_id"
```
Comment (delarocamcas) [inline Grouper.py:478]: this has been moved to the master function

File: Grouper.py [hunk 6]
```diff
@@ -540,70 +554,93 @@ def dynamic_threshold(token1, token2, base_threshold=75, max_threshold=90):
     else:
         return base_threshold + ((avg_len - 5) / 10) * (max_threshold - base_threshold)
 
-township_pattern = r'^[trs]\d{1,3}[nsew]?$'
```
Comment (delarocamcas) [inline Grouper.py:556]: Even though it looks like I changed alot in this section, I just packaged it into functions. I checked line by line to make sure every operation is accounted for.

File: Grouper.py [hunk 7]
```diff
@@ -615,200 +652,192 @@ def apply_aliases(text, alias_map):
...
-grouped['normalized_locality'] = grouped['normalized_locality'].apply(lambda t: apply_aliases(t, merged))
+# --- Cosine similarity ---
+def group_by_similarity(grouped, id_matrix):
+    """
+    Assign Suggested_ID and Grouper_ID using cosine similarity groupings.
+    """
+    similarity = cosine_similarity(id_matrix)
 
-# --- Rebuild TF-IDF matrix on alias-applied text ---
-X = vectorizer.fit_transform(grouped['normalized_locality'])
```
Comment (delarocamcas) [inline Grouper.py:622]: generally not ideal practice to use math variables like X, Y, Z as variables , so I renamed this to id_matrix

File: Grouper.py [hunk 8]
```diff
@@ -615,200 +652,192 @@ def apply_aliases(text, alias_map):
...
-columns_to_export = [col for col in columns_to_export if col in grouped.columns]
+def reorder_similar_singletons(grouped, similarity, min_similarity=0.80):
+    """
+    Reorder singletons based on similarity to closest larger group.
+    Identify singleton groups and place them after the most similar non-singleton group.
+    """
+    start_time = time.time()
+    print("Identifying singleton placements...")
 
-import re
```
Comment (delarocamcas) [inline Grouper.py:809]: moved all imports to top

File: Grouper.py [hunk 9]
```diff
@@ -615,200 +652,192 @@ def apply_aliases(text, alias_map):
...
+    # Now define non-singleton_ids AFTER filtering valid singleton_ids
+    non_singleton_ids = [gid for gid in group_id_to_indices if gid not in singleton_ids]
+
+    singleton_inserts = {}
+    for singleton_id in singleton_ids:
+        singleton_idx = group_id_to_indices[singleton_id][0]
+        best_score = -1
+        best_match_id = None
+
+        for gid in non_singleton_ids:
```
Comment (delarocamcas) [inline Grouper.py:820]: generally a forloop inside a forloop isn't ideal. I decided to not make any effeciency improvements/ functional changes to the code yet. This can be addressed in future smaller and limited PRs.

File: Grouper.py [hunk 10]
```diff
@@ -827,9 +859,84 @@ def grouper_sort_key(gid):
...
+        by='Grouper_ID',
+        key=lambda col: col.map(lambda gid: grouper_sort_key(gid, singleton_inserts))
+    )
+
+
+    output_file = os.path.splitext(csv_path)[0] + '-key.csv'
+    export_df.to_csv(output_file, index=False, encoding='utf-8-sig')
+    print(f"Exported with suggested groups to: {output_file}")
+
+def grouper_main():
```
Comment (delarocamcas) [inline Grouper.py:899]: master function, includes all methods in order , commented for easy comprehension of major steps.

**Review by Cmeyer56555 — APPROVED** [review]
